File: src/lib/polycam-normalizer.ts
type Dict = Record<string, unknown>;

const DEFAULT_WALL_HEIGHT_IN = 96;

function asDict(value: unknown): Dict {
  return value && typeof value === 'object' ? (value as Dict) : {};
}

function asArray(value: unknown): Dict[] {
  return Array.isArray(value) ? value.map(asDict) : [];
}

function blocking(code: string, message: string, path: string) {
  return { code, message, severity: 'blocking', path };
}

export function normalizePolycamToCanonical(metaInput: unknown, planInput: unknown): Dict {
  const meta = asDict(metaInput);
  const model = structuredClone(asDict(planInput));

  const project = asDict(model.project);
  project.project_id = project.project_id || meta.project_id || 'proj_unknown';
  project.capture_source = 'polycam';
  project.units = 'imperial';
  model.project = project;

  const errors: Dict[] = [];

  const entities = asDict(model.entities);
  const walls = asArray(entities.walls);
  const openings = asArray(entities.openings);
  const rooms = asArray(entities.rooms);

  walls.forEach((wall, i) => {
    if (wall.height_in === undefined || wall.height_in === null || wall.height_in === '') {
      wall.height_in = DEFAULT_WALL_HEIGHT_IN;
    }
    if (!(Number(wall.height_in) > 0)) {
      errors.push(blocking('E-GEO-006', 'Wall height must be > 0', `/entities/walls/${i}/height_in`));
    }
  });

  openings.forEach((opening, i) => {
    if (opening.type === 'door') {
      if (opening.sill_in === undefined || opening.sill_in === null) opening.sill_in = 0;
    } else if (opening.type === 'window') {
      if (opening.sill_in === undefined || opening.sill_in === null) {
        errors.push(
          blocking('E-GEO-005', 'Window opening requires sill_in for permit track', `/entities/openings/${i}/sill_in`)
        );
      }
    }
  });

  rooms.forEach((room, i) => {
    if (!String(room.name || '').trim()) {
      errors.push(blocking('E-META-003', 'Room name is required for permit', `/entities/rooms/${i}/name`));
    }
  });

  entities.walls = walls;
  entities.openings = openings;
  entities.rooms = rooms;
  model.entities = entities;

  model.schema_version = model.schema_version || '1.0.0';

  const revisions = Array.isArray(model.revisions) ? model.revisions : [];
  if (revisions.length === 0) {
    revisions.push({
      rev_id: 'rev_normalized',
      timestamp: new Date().toISOString(),
      actor: 'system',
      source: 'intake',
      operations: [],
      model_hash: 'normalized_seed',
    });
  }
  model.revisions = revisions;

  const validation = asDict(model.validation);
  const existingWarnings = Array.isArray(validation.warnings) ? validation.warnings : [];
  const parserConfidence = Number.isFinite(validation.parser_confidence)
    ? Number(validation.parser_confidence)
    : 0.9;

  model.validation = {
    status: errors.length ? 'invalid' : 'valid',
    errors,
    warnings: existingWarnings,
    parser_confidence: parserConfidence,
  };

  return model;
}

File: scripts/polycam_normalizer.py
from __future__ import annotations

from copy import deepcopy
from datetime import datetime, timezone
from typing import Any

DEFAULT_WALL_HEIGHT_IN = 96


def _now_iso() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def _blocking(code: str, message: str, path: str) -> dict[str, Any]:
    return {"code": code, "message": message, "severity": "blocking", "path": path}


def normalize_polycam_to_canonical(meta: dict[str, Any], plan: dict[str, Any]) -> dict[str, Any]:
    model = deepcopy(plan)

    project = model.setdefault("project", {})
    project["project_id"] = project.get("project_id") or meta.get("project_id", "proj_unknown")
    project["capture_source"] = "polycam"
    project["units"] = "imperial"

    errors: list[dict[str, Any]] = []

    entities = model.setdefault("entities", {})
    walls = entities.setdefault("walls", [])
    openings = entities.setdefault("openings", [])
    rooms = entities.setdefault("rooms", [])

    for i, wall in enumerate(walls):
        if wall.get("height_in") in (None, ""):
            wall["height_in"] = DEFAULT_WALL_HEIGHT_IN
        if wall.get("height_in", 0) <= 0:
            errors.append(_blocking("E-GEO-006", "Wall height must be > 0", f"/entities/walls/{i}/height_in"))

    for i, opening in enumerate(openings):
        kind = opening.get("type")
        if kind == "door":
            opening["sill_in"] = opening.get("sill_in", 0)
        elif kind == "window":
            if opening.get("sill_in") is None:
                errors.append(
                    _blocking(
                        "E-GEO-005",
                        "Window opening requires sill_in for permit track",
                        f"/entities/openings/{i}/sill_in",
                    )
                )

    for i, room in enumerate(rooms):
        if not str(room.get("name", "")).strip():
            errors.append(_blocking("E-META-003", "Room name is required for permit", f"/entities/rooms/{i}/name"))

    model.setdefault("schema_version", "1.0.0")
    model.setdefault("revisions", [])
    if not model["revisions"]:
        model["revisions"].append(
            {
                "rev_id": "rev_normalized",
                "timestamp": _now_iso(),
                "actor": "system",
                "source": "intake",
                "operations": [],
                "model_hash": "normalized_seed",
            }
        )

    model["validation"] = {
        "status": "invalid" if errors else "valid",
        "errors": errors,
        "warnings": model.get("validation", {}).get("warnings", []),
        "parser_confidence": model.get("validation", {}).get("parser_confidence", 0.9),
    }

    return model
